### ISSUE TITLE
Allow changing the default xrootd user

### ISSUE THREAD

**Issue by hseverini** [xrootd/xrootd#2346]
Hi Guilherme,

as promised, here the github issue. We'd love to be able to change the default xrootd service user to another user, like usatlas1 in our case, which should make it easier for a single-VO ATLAS xrootd service like ours to interact with the underlying cephfs file system, such that jobs running on the compute nodes can write into the same directory tree as the xrootd service on the DTN.

Quoting you for reference: "The solution will likely involve using an environment file for the systemd unit, and setting the user to be used for XRootD there rather than hard-code it the way it's done now."

I'll be happy to be a guinea pig and test this!

Thanks very much,
      Horst

**Comment by amadio** [comment]
Dear @hseverini, I have good news for you. Actually, this doesn't need any change in XRootD itself. If you'd like to change the user running the XRootD service, you can create a directory named `xrootd@.service.d` within `/etc/systemd/system`, and create a file `user.conf` (the name doesn't really matter) inside it, with the following contents:
```
[Service]
User=usatlas1
```
This will override the `User=xrootd` coming from the standard systemd unit we install. If you want it to only have an effect for a particular XRootD server (i.e. if you run it as xrootd@usaltas1 using `/etc/xrootd/xrootd-usatlas1.conf` as config), you can name the directory `xrootd@usatlas1.service.d` and the change will only take effect for that particular configuration file.

Please let us know if this is enough to solve this issue so that we can close this ticket.

Best regards,
-Guilherme

**Comment by abh3** [comment]
Let's put this explanation in the "how to" notes pointed to by the documentation page.

**Comment by hseverini** [comment]
Thanks Guilherme!

So on this new xrootd server, I think we will only be running one xrootd 
service, namely ***@***.***, started from the config file 
/etc/xrootd/xrootd-cephfs.cfg .

So do I understand this correctly that I'd need to create a directory

***@***.***/
or
***@***.***/

and put this user.conf file in there then?

Andy, which one of these two options would you suggest for our setup?

Thanks a lot,

	Horst

**Comment by abh3** [comment]
I'll defer to he release manager on this one but you should notice that the two choices presented above are identical. So, it seems the choice has been made a priori.

**Comment by hseverini** [comment]
Dear Horst,

I'm replying by email since GitHub redacted some of the text thinking it
was a password.

On Thu, Oct 03, 2024 at 01:05:21AM +0200, Horst Severini wrote:
> So on this new xrootd server, I think we will only be running one xrootd 
> service, namely ***@***.***, started from the config file 
> /etc/xrootd/xrootd-cephfs.cfg .
> 
> So do I understand this correctly that I'd need to create a directory
> 
> ***@***.***/
> or
> ***@***.***/
> 
> and put this user.conf file in there then?

Yes, that's correct. You can override anything from the standard unit
in files inside this directory too, overriding the `User=` config was
just an example. See the EXAMPLES section in `man 5 systemd.unit` for
more information.

> Andy, which one of these two options would you suggest for our setup?

The two options above both work and should be equivalent if you only
have one daemon running. If you want to run more than one daemon, then
the second option is best, as it only affects ***@***.*** and
not the other daemons, but that's the only difference. The important
thing is to not need to edit the systemd unit that is installed by
XRootD itself, and both options work for that.

Best regards,
-Guilherme

**Comment by hseverini** [comment]
Hi Guilherme, Andy,

I tried this, but it doesn't quite seem to work:

----
***@***.*** ~]# ls -l ***@***.***/
total 4
-rw-r--r--.  1 root root   24 Oct  3 17:02 user.conf

***@***.*** ~]# cat ***@***.***/user.conf
[Service]
User=usatlas1

***@***.*** ~]# id usatlas1
uid=1216(usatlas1) gid=978(xrootd) groups=978(xrootd)

***@***.*** ~]# systemctl start ***@***.***
Warning: The unit file, source configuration file or drop-ins of 
***@***.*** changed on disk. Run 'systemctl daemon-reload' to 
reload units.

***@***.*** ~]# ps auxw | grep xroot
xrootd     99150  0.5  0.0 330100 22540 ?        Ssl  17:12   0:00 
/usr/bin/xrootd -l /var/log/xrootd/xrootd.log -c 
/etc/xrootd/xrootd-cephfs.cfg -k fifo -s /run/xrootd/xrootd-cephfs.pid 
-n cephfs
----

So the user of the service wasn't changed from xrootd to usatlas1.

I did run 'systemctl daemon-reload' several times, but that didn't seem 
to help, I get the same warning every time I start or stop the service.

I also tried to move this user.conf file from ***@***.***/ 
to ***@***.***/, but that didn't make a difference, same warning.

What should I try next?

Thanks a lot,

	Horst

**Comment by amadio** [comment]
Dear @hseverini,

Could you please paste here the output of the command below?
```sh
systemctl show xrootd@cephfs | grep User
```
This can be used to check which user is in effect. Please make sure to run `systemctl daemon-reload` before running this command and that the unit which was installed by XRootD's RPM package has not been modified afterwards.

Cheers,

**Comment by hseverini** [comment]
Hi Guilherme,

sure:

----
***@***.*** ~]# systemctl show ***@***.*** | grep User
User=xrootd
DynamicUser=no
PrivateUsers=no
----

So somehow while the system can see this user.conf file, because it 
throws that warning, it doesn't really 'honor' it.

Where can we look for other clues on how to fix this?

Thanks,

	Horst

**Comment by wyang007** [comment]
I just gave this a try, and it seems to work for me. I will try a few more cases. This solve a big problem. 

Can we keep this ticket open for a while? I think another issue related to enabling non-xrootd users to run is the permission of /var/spool/xrootd and a few other directories like that. On our system, we have several instances of xrootd run by different users. The issue is:
1.  the directory /var/spool/xrootd (and a few others) is owned by user xrootd. 
2. we need to manually create /var/spool/xrootd/<instance> if <instance> is not run by user xrootd. This is a one time task so it is OK.
3. when updating rpms, /var/spool/xrootd's permission is set to 700. this prevent <instance> from starting

So, can the rpm script not changing /var/spool/xrootd if it exists?

**Comment by hseverini** [comment]
Hi Wei,

hmm, how come this works for you but not for me? What O/S and version 
are you running? You didn't get the unit file warnings I did? Did you 
need to make any other changes?

Where did you add this config file in ***@***.***/ or in 
***@***.***/ ?

And what does 'systemctl show ***@***.*** | grep User' show for you?

Thanks,

	Horst

**Comment by wyang007** [comment]
I use xrootd at instance_name dot service dot d. I put a file local.conf in side the directory. It has
[Service]
User=usatlas1
Group=abc

These change shows up in "systemctl show xrootd at instance_name"

**Comment by hseverini** [comment]
Hi Wei,

 > I use xrootd at instance_name dot service dot d. I put a file
 > local.conf in side the directory. It has
 > [Service]
 > User=usatlas1
 > Group=abc
 >
 > These change shows up in "systemctl show xrootd at instance_name"

That's odd. Can you show us an 'ls -l' of these files and directories, 
to see whether mine somehow have incorrect permissions or ownership?

What does the 'Group' line do? I don't have that, and I don't know what 
it should do. Or is that just the unix group?

Also, which O/S version is your server running? Also Alma9, or some 
other EL9 or EL8 version? Not sure if that makes a difference, but worth 
checking ...

Thanks a lot,

	Horst

**Comment by wyang007** [comment]
This is what I have:
$ ls -l /etc/systemd/system/xrootd@rubinsummit.service.d/
total 4
-rw-r--r-- 1 root root 33 Oct  6 22:28 local.conf
$ cat /etc/systemd/system/xrootd@rubinsummit.service.d/local.conf 
[Service]
User=lsstsvc1
Group=gu

This works on EL8. "Group" is unix group.

**Comment by hseverini** [comment]
Hi Wei,

hmm, that looks the same as what I have. Maybe there's a difference in 
security or permissions between el8 and el9?

Or do you have selinux turned off? We have that turned on; maybe that's 
what's preventing this from working?

Thanks,
good night,

	Horst

**Comment by wyang007** [comment]
I believe we set selinux off on all of our scientific computing systems.

**Comment by hseverini** [comment]
Hi Wei,

okay, I disabled selinux, and then it worked! Here are the steps I took 
to get it running:

----
setenforce 0
chown -R usatlas1. /etc/grid-security/xrd/
chown -R usatlas1. /var/log/xrootd /var/run/xrootd /var/spool/xrootd
chown usatlas1. /etc/xrootd/macaroon-secret
systemctl daemon-reload
----

Then it started without errors, and some TPC test transfers worked fine.

Can you please try some test transfers from your end as well?
The endpoint again is

https://se0.oscer.ou.edu:9090/ourdisk/hpc/xrd_test/

Thanks a lot,

	Horst

**Comment by abh3** [comment]
The question is whether you can turn selinux on after starting xrootd. Then it follows how do you turn it off before starting xrootd should a restart occur. I'm not sure you really want to run in permissive mode.

**Comment by hseverini** [comment]
Hi Andy, Guilherme,

right, ideally you'd want this to work with selinux turned on.

Guilherme, do you know of an easy way to allow these changes in selinux? 
I'm assuming you already have other selinux setting in your build, so 
hopefully you can add this one there as well?

Thanks a lot,

	Horst

**Comment by wyang007** [comment]
TPC is still working between SLAC and OU

**Comment by hseverini** [comment]
Glad to hear that!

So now we just need to get the selinux stuff fixed in the RPM distro, 
and then we're in business! :)

Thanks a lot,

	Horst

**Comment by hseverini** [comment]
Hi again,

here again the question about selinux, to make sure it's logged properly 
in the ticket:

I had to set selinux to permissive on reboot, since without that, xrootd 
still won't start with this change to the usatlas1 user. What's the 
status of adding the appropriate selinux configuration in the xrootd 
rpms, so that you can change the user with selinux turned on?

I'm assuming this needs to go into the xrootd-selinux RPM? Who's in 
charge of that? :)

Thanks a lot,

	Horst

**Comment by amadio** [comment]
I will see how to do this and fix it in the next release. Cheers,

**Comment by hseverini** [comment]
Sounds great, thanks Guilherme!

	Horst

**Comment by amadio** [comment]
I have tried [this](https://github.com/xrootd/xrootd/issues/2346#issuecomment-2387920130) with SELinux enabled, and everything worked. @hseverini I think there may be something special (like extra enabled SELinux modules) in your environment that makes it fail. I will drop this from 5.8 release milestone, as I'd like to release soon and I don't have anything concrete to adress (other than the documentation for this, which can be done after we really understand the problem).

Could you please enable SELinux, try again customizing the user, then send us the output of `ausearch -ts recent -m avc` after it fails? If you can do the investigation of why exactly it fails, I can reproduce the problem and suggest how to fix. For now, all I can say is that I had it working with SELinux on my machine. Thank you!

**Comment by hseverini** [comment]
Hi Guilherme,

okay, I just enabled selinux again and rebooted, and it failed again as 
before, but your command isn't reporting anything:

----
***@***.*** ~]# systemctl restart ***@***.***
Warning: The unit file, source configuration file or drop-ins of 
***@***.*** changed on disk. Run 'systemctl daemon-reload' to 
reload units.
***@***.*** ~]# ps auxw | grep xroot
root        6926  0.0  0.0 221660  2048 pts/0    S+   11:37   0:00 grep 
--color=auto xroot
***@***.*** ~]# ausearch -ts recent -m avc
<no matches>
----

I tried 'systemctl daemon-reload' and it doesn't report any errors, but 
it doesn't fix anything, either.

I'm not aware of any special configuration we have here which would 
interfere with selinux, but I know very little about selinux, so I 
wouldn't know what to look for here.

Anything else I can try before turning selinux back off and rebooting 
again, to make it work again for now?

Thanks,
	
	Horst

**Comment by panta-123** [comment]
We are interested in this. 
Our xrootd version is v5.8.4 , can this be tested with v5.8.4 version ?

**Comment by amadio** [comment]
In principle, there is nothing preventing you from changing settings and using a different user with XRootD and selinux enabled. I have done it on Alma 9 without problems. What may prevent you is some local setup that is unexpected to selinux and missing selinux context on files served by XRootD, we'd need to take a specific case and debug to see what could cause problems and suggest a checklist to go through to set it up. This is what this issue is about. If you also have problems with selinux, let us know and together with @hseverini we debug it.

**Comment by amadio** [comment]
@hseverini I had it working with selinux on my own Alma 9 test server. In order to be able to debug why it doesn't work in your setup, we need additional information. For example, checking what [context](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/7/html/selinux_users_and_administrators_guide/sect-security-enhanced_linux-working_with_selinux-selinux_contexts_labeling_files) your files have, the error you get when XRootD fails to start, which permission is not granted that would be required for XRootD to be able to start, etc.

**Comment by amadio** [comment]
Hi @hseverini Could you please send to me by email a full log of your server that does not start when SELinux is enabled? Please enable all tracing you can. Also, could you please also try to start a server with the same configuration, but using a local filesystem instead of a cephfs mount as `all.export` directory? We need to narrow down where the problem lies. Please also try `ls -Z /path/to/cephfs` and send us the output as well. Thank you.

**Comment by hseverini** [comment]
Yes, let's by-pass github, I'm not fond of their email interface.

The xrootd instance on both se0 and se2 is called cephfs:

***@***.*** ~]# ls -al /etc/xrootd/xrootd-cephfs.cfg
-rw-r--r--. 1 root root 2393 Aug 12 15:40 /etc/xrootd/xrootd-cephfs.cfg

***@***.*** ~]# ls -al /etc/xrootd/xrootd-cephfs.cfg
-rw-r--r--. 1 root root 2393 Aug 12 15:40 /etc/xrootd/xrootd-cephfs.cfg

And both instances are setup identically as far as I can tell,
same configuration.

And the other file is

***@***.***/user.conf

And 'systemctl status ***@***.***' (and start/stop) 'works' on both, 
but it fails on se2, because the change of ownership is ignored 
for some reason. And I don't understand why, since both config files
are identical.

So what's different?

Thanks,

	Horst
Thanks,

	Horst
